Backend plugins: Support Forward OAuth Identity for backend data source plugins

billoley opened the pull request:
fixes #26023 

If configured to do so, pass the oauth token to the backend datasource code.
This is already being done in ds_proxy.go for backend proxied datasources.
This change moves the code from the addOAuthPassThruAuth method to a new package oauthtoken with slight refactoring.
This allows both datasource_plugin_wrapper_v2.go and ds_proxy.go to use the same code to get an up-to-date token.

billoley commented:
Code flow through social.GetCurrentOAuthToken is already tested by ds_proxy_test.go when getDirector() is called with oauth configured

billoley commented:
Should this PR be rebased on branch v7.1.x instead of master.... which will be 7.2.0?

kminehart commented on pkg/login/social/social.go:
```diff
@@ -225,3 +233,92 @@ var GetOAuthProviders = func(cfg *setting.Cfg) map[string]bool {
 	return result
 }
+
+var GetOAuthHttpClient = func(name string) (*http.Client, error) {
```
```suggestion
func GetOAuthHttpClient(name string) (*http.Client, error) {
```

kminehart commented on pkg/login/social/social.go:
```diff
@@ -225,3 +233,92 @@ var GetOAuthProviders = func(cfg *setting.Cfg) map[string]bool {
+var GetOAuthHttpClient = func(name string) (*http.Client, error) {
+	name = strings.TrimPrefix(name, "oauth_")
+
+	// handle call back
+	tr := &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: setting.OAuthService.OAuthInfos[name].TlsSkipVerify,
+		},
+	}
```
```suggestion
	info, ok := setting.OAuthService.OAuthInfos[name]
	if !ok {
		return nil, fmt.Errorf("Could not find %s in OAuth Settings", name)
	}
	
	// handle call back
	tr := &http.Transport{
		Proxy: http.ProxyFromEnvironment,
		TLSClientConfig: &tls.Config{
			InsecureSkipVerify: info.TlsSkipVerify,
		},
	}
```

If `name` is not in the `OAuthInfos` map, trying to access `TlsSkipVerify` will panic.

kminehart commented on pkg/login/social/social.go:
```diff
@@ -225,3 +233,92 @@ var GetOAuthProviders = func(cfg *setting.Cfg) map[string]bool {
+		Transport: tr,
+	}
+
+	if setting.OAuthService.OAuthInfos[name].TlsClientCert != "" || setting.OAuthService.OAuthInfos[name].TlsClientKey != "" {
+		cert, err := tls.LoadX509KeyPair(setting.OAuthService.OAuthInfos[name].TlsClientCert, setting.OAuthService.OAuthInfos[name].TlsClientKey)
```
```suggestion
	if info.TlsClientCert != "" || info.TlsClientKey != "" {
		cert, err := tls.LoadX509KeyPair(info.TlsClientCert, info.TlsClientKey)
```

kminehart commented on pkg/login/social/social.go:
```diff
@@ -225,3 +233,92 @@ var GetOAuthProviders = func(cfg *setting.Cfg) map[string]bool {
+		tr.TLSClientConfig.Certificates = append(tr.TLSClientConfig.Certificates, cert)
+	}
+
+	if setting.OAuthService.OAuthInfos[name].TlsClientCa != "" {
+		caCert, err := ioutil.ReadFile(setting.OAuthService.OAuthInfos[name].TlsClientCa)
```
```suggestion
	if info.TlsClientCa != "" {
		caCert, err := ioutil.ReadFile(info.TlsClientCa)
```

kminehart commented on pkg/login/social/social.go:
```diff
@@ -225,3 +233,92 @@ var GetOAuthProviders = func(cfg *setting.Cfg) map[string]bool {
+	return oauthClient, nil
+}
+
+var GetCurrentOAuthToken = func(ctx context.Context, user models.SignedInUser) (*oauth2.Token, error) {
```
```suggestion
func GetCurrentOAuthToken(ctx context.Context, user models.SignedInUser) (*oauth2.Token, error) {
```

kminehart commented on pkg/login/social/social.go:
```diff
@@ -225,3 +233,92 @@ var GetOAuthProviders = func(cfg *setting.Cfg) map[string]bool {
+}
+
+var GetCurrentOAuthToken = func(ctx context.Context, user models.SignedInUser) (*oauth2.Token, error) {
+	authInfoQuery := &models.GetAuthInfoQuery{UserId: user.UserId}
```
This tells me that you did not run `go fmt` :P

billoley replied:
I thought I ran the required linting and formatting.  I will take a look.

kminehart commented on pkg/plugins/datasource/wrapper/datasource_plugin_wrapper_v2.go:
```diff
@@ -51,6 +53,23 @@ func (tw *DatasourcePluginWrapperV2) Query(ctx context.Context, ds *models.DataS
+			token, err := social.GetCurrentOAuthToken(ctx, *query.User)
+			if err != nil {
+				tw.logger.Error("Error fetching oauth information for user", "error", err)
+			} else if token != nil {
+				delete(query.Headers, "Authorization")
+				query.Headers["Authorization"] = fmt.Sprintf("%s %s", token.Type(), token.AccessToken)
+			}
```
```suggestion
			}
			delete(query.Headers, "Authorization")
			query.Headers["Authorization"] = fmt.Sprintf("%s %s", token.Type(), token.AccessToken)
```

I think in most cases, and `social.GetCurrentOauthToken` included, it's fair to assume that if `err` is nil, then the other value in the response is not.

kminehart commented on pkg/api/pluginproxy/ds_proxy.go:
```diff
@@ -211,7 +208,15 @@ func (proxy *DataSourceProxy) getDirector() func(req *http.Request) {
+				logger.Error("Error fetching oauth token for user", "error", err)
+				return
+			}
+			if token != nil {
+				req.Header.Del("Authorization")
+				req.Header.Add("Authorization", fmt.Sprintf("%s %s", token.Type(), token.AccessToken))
+			}
```
If `err` is nil then it's very fair to assume that `token` is not.

```suggestion
			req.Header.Del("Authorization")
			req.Header.Add("Authorization", fmt.Sprintf("%s %s", token.Type(), token.AccessToken))
```

kminehart reviewed (dismissed):
I requested some minor style changes. I am much less familiar with Grafana than the listed reviewers so I would still like them to review this.

This would be super useful for users of the github plugin (https://github.com/grafana/grafana-github-datasource)

kminehart commented:
It would be nice if, instead of configuring this per OAuth provider, we could configure this per datasource.

billoley commented:
> It would be nice if, instead of configuring this per OAuth provider, we could configure this per datasource.

Passing the oauth token through _is_ configured on a per datasource basis for both backand plugins and for the proxied plugins

Backend
https://github.com/grafana/grafana/pull/27055/files#diff-71d6e3ee71ac498cfe8fcd0e3ee9a965R60

Proxied
https://github.com/grafana/grafana/pull/27055/files#diff-cd5947c41782508115b4ac9e25218aa1R210

kminehart commented:
Oh! For some reason I assumed that it was per datasource. Sorry about that!

billoley commented:
@kminehart - I'm pretty sure that I made all requested changes and then rebased on master to get the ci build to complete.  One of your requested change (that I think I made) is blocking the PR for some reason.

Also - Any idea if someone can review this PR to get it into 7.2?

marefr reviewed (changes requested):
Thanks. In general looks good. There's a few minor things I would like to see changed, see comments.

The biggest problem right now is that the social package have to much responsibility since it now besides handling OAuth providers and provide OAuth mechanism it also makes database calls. 

To me it seems like we're lacking a package that could coordinate receiving of stored OAuth token, refresh it if needed and store the updated token in the database. This package could then be used by ds_proxy.go, datasource_plugin_wrapper_v2.go  and potentially other code since backend plugins also has support for resources and potentially other capabilities in the future.

Question is just where to put this package and how to name it. My suggestions that comes to mind:
- pkg/services/oauth
- pkg/services/oauthtoken

Thoughts/questions?

billoley commented:
> Thanks. In general looks good. There's a few minor things I would like to see changed, see comments.
> 
> The biggest problem right now is that the social package have to much responsibility since it now besides handling OAuth providers and provide OAuth mechanism it also makes database calls.
> 
> To me it seems like we're lacking a package that could coordinate receiving of stored OAuth token, refresh it if needed and store the updated token in the database. This package could then be used by ds_proxy.go, datasource_plugin_wrapper_v2.go and potentially other code since backend plugins also has support for resources and potentially other capabilities in the future.
> 
> Question is just where to put this package and how to name it. My suggestions that comes to mind:
> 
> * pkg/services/oauth
> * pkg/services/oauthtoken
> 
> Thoughts/questions?

I chose the social package as it seemed most the most appropriate from existing packages, but figured that the code owners might have other ideas.  I think either suggestion would be fine, but probably oauthtoken to keep it separate from the Grafana code that implements an oauth provider.

billoley commented on pkg/login/social/social.go:
```diff
@@ -225,3 +233,97 @@ var GetOAuthProviders = func(cfg *setting.Cfg) map[string]bool {
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create http client for oauth operation error:%s", err)
+	}
+	ctx = context.WithValue(ctx, oauth2.HTTPClient, client)
```
I saw an http ssl error on getting a new token (refresh) and then found that login_oauth.go was setting the http client into the context.  It made sense that the same thing should be done for refreshing the token.

billoley replied:
Moved the OAuth HTTP Client changes to a new PR -> https://github.com/grafana/grafana/pull/27523
It will be a lot easier to include that new call on this PR once that PR is approved and merged.
I will back that code out of this PR (except for calling it once it's merged in)

marefr replied:
Great. I have reviewed #27523

billoley commented on pkg/plugins/datasource/wrapper/datasource_plugin_wrapper_v2.go:
```diff
@@ -51,6 +53,22 @@ func (tw *DatasourcePluginWrapperV2) Query(ctx context.Context, ds *models.DataS
+		query.Headers = make(map[string]string)
+	}
+
+	if ds.JsonData != nil && ds.JsonData.Get("oauthPassThru").MustBool() {
```
I think this helper method would fit well in pkg/services/oauthtoken since both datasource_plugin_wrapper_v2.go and ds_proxy.go will be accessing the same package to get the current OAuthToken.

marefr replied:
Yeah makes sense. I wrote that comment before realizing that things probably should be moved around.

billoley commented on pkg/login/social/social.go:
```diff
@@ -225,3 +233,97 @@ var GetOAuthProviders = func(cfg *setting.Cfg) map[string]bool {
+
+var GetCurrentOAuthToken = func(ctx context.Context, user models.SignedInUser) (*oauth2.Token, error) {
+	authInfoQuery := &models.GetAuthInfoQuery{UserId: user.UserId}
+	if err := bus.Dispatch(authInfoQuery); err != nil {
```
Would the same code be fine if it is in the new oauthtoken package as you suggested?

If GetCurrentOAuthToken (in oauthtoken, not social) has to take this as an argument, then we will have the database code twice -- once in datasource_plugin_wrapper_v2.go and once in ds_proxy.go.

marefr replied:
Yes, parts of it. Would make sense to have database interactions in oauthtoken package. oauthtoken package should use the social package for retrieving the token (and refresh if needed) since the social package has all the internal knowledge of providers, http clients etc. Basically the oauthtoken package will be the coordinator for getting a token, but reach out to other packages for getting the user token from database, refresh it, if needed, and store it in database if refreshed. 

Does this make sense?

billoley committed:
Propagate OAuth token to backend datasource plugins, refactor fetching of current token

More

More

billoley commented:
Sorry for the force push.  Between splitting the PR and rebasing a bunch of prior commits, this seemed cleaner (and easier)

billoley committed:
OAuth token will not be available is user logged in another way

marefr commented:
Thanks. Planning to review this again soon, hopefully today or else beginning next week. Have had a lot of other things going on.

billoley commented on pkg/login/social/social.go:
```diff
@@ -258,8 +258,8 @@ func GetOAuthHttpClient(name string) (*http.Client, error) {
 
 	if info.TlsClientCert != "" || info.TlsClientKey != "" {
 		cert, err := tls.LoadX509KeyPair(info.TlsClientCert, info.TlsClientKey)
```
Moved this fix to another PR https://github.com/grafana/grafana/pull/27804 since it is a bug instead of an improvement related to this task.

billoley commented:
> Thanks. Planning to review this again soon, hopefully today or else beginning next week. Have had a lot of other things going on.

Understand.  Thanks for your time.

marefr commented:
Sorry for keeping you waiting. Will make sure to review it beginning next week.

billoley commented:
Any update on review?

aknuds1 commented:
We're supposed to review it tomorrow.

marefr commented:
Reviewing this now. Really sorry for keeping you waiting @billoley

marefr reviewed (changes requested):
Great work. Some minor things left related to logging/error handling.

billoley commented on pkg/plugins/datasource/wrapper/datasource_plugin_wrapper_v2.go:
```diff
@@ -51,6 +53,27 @@ func (tw *DatasourcePluginWrapperV2) Query(ctx context.Context, ds *models.DataS
+
+	if oauthtoken.IsOAuthPassThruEnabled(ds) {
+		// skip AddOAuthPassThruAuth for alerts
+		if query.User != nil {
```
I agree that moving the user nil check into oauthtoken.GetCurrentOAuthToken makes sense.  If there is no user, then there is no token and we can return nil, nil.

I disagree with moving the IsOAuthPassThruEnabled(ds) check into GetCurrentOAuthToken.  As written, GetCurrentOAuthToken returns the token (possibly refreshed) if there is one.  The check on whether IsOAuthPassThruEnabled(ds) is true or false does not fit into the gereralized purpose of GetCurrentOAuthToken.  It seems wrong to me unless there is something I am missing.

However.... your code.   Let me know.

marefr replied:
I was only referring to the nil check and not moving `IsOAuthPassThruEnabled`

aknuds1 commented on pkg/services/oauthtoken/oauth_token_util.go:
```diff
@@ -0,0 +1,77 @@
+	authProvider := authInfoQuery.Result.AuthModule
+	connect, err := social.GetConnector(authProvider)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get OAuth connector error=%s", err)
```
```suggestion
		return nil, fmt.Errorf("failed to get OAuth connector error=%s", err)
```

aknuds1 commented on pkg/services/oauthtoken/oauth_token_util.go:
```diff
@@ -0,0 +1,77 @@
+
+	client, err := social.GetOAuthHttpClient(authProvider)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create OAuth http client error=%s", err)
```
```suggestion
		return nil, fmt.Errorf("failed to create OAuth http client error=%s", err)
```

aknuds1 commented on pkg/services/oauthtoken/oauth_token_util.go:
```diff
@@ -0,0 +1,77 @@
+			OAuthToken: token,
+		}
+		if err := bus.Dispatch(updateAuthCommand); err != nil {
+			return nil, fmt.Errorf("Failed to update auth info during token refresh userId=%d username=%s error=%s", user.UserId, user.Login, err)
```
```suggestion
			return nil, fmt.Errorf("failed to update auth info during token refresh, userId=%d username=%s error=%s", user.UserId, user.Login, err)
```

aknuds1 reviewed (commented):
Generally LGTM, but spotted some nits.

billoley committed:
Apply suggestions from code review

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 commented on pkg/services/oauthtoken/oauth_token_util.go:
```diff
@@ -0,0 +1,77 @@
+	// TokenSource handles refreshing the token if it has expired
+	token, err := connect.TokenSource(ctx, persistedToken).Token()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to retrieve access token from OAuth provider=%s userid=%d username=%s error=%s", authInfoQuery.Result.AuthModule, user.UserId, user.Login, err)
```
```suggestion
		return nil, fmt.Errorf("failed to retrieve access token from OAuth, provider=%s userid=%d username=%s error=%s", authInfoQuery.Result.AuthModule, user.UserId, user.Login, err)
```
```suggestion
		return nil, fmt.Errorf("Failed to retrieve access token from OAuth provider=%s userid=%d username=%s error=%s", authInfoQuery.Result.AuthModule, user.UserId, user.Login, err)
```
```suggestion
		return nil, fmt.Errorf("Failed to retrieve access token from OAuth provider=%s userid=%d username=%s error=%s", authInfoQuery.Result.AuthModule, user.UserId, user.Login, err)
```

aknuds1 reviewed (commented):
Spotted a remaining nit.

billoley committed:
Update pkg/services/oauthtoken/oauth_token_util.go

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

billoley commented on pkg/services/oauthtoken/oauth_token_util.go:
```diff
@@ -0,0 +1,77 @@
+	if err := bus.Dispatch(authInfoQuery); err != nil {
+		logger.Debug("No oauth token for user", "userid", user.UserId, "username", user.Login)
+		// Not necessarily an error.  User may be logged in another way.
+		return nil, nil
```
At first, I was returning an error here, but found that when I was logged in as admin with username/password, the Dispatch call returned an error.  Should that be true or am I mistaken?  

If Dispatch returning an error in the non-oauth case is correct, then I think this should remain as-is returning nil, nil instead of propagating the error.

marefr replied:
Returning `nil, nil` I don't like. As a consumer and Go developer I'm used to checking the err for non-nil and if not expect other returned value is valid. With this you have to check both return values for nil. But I understand why that may be needed here. Returning a special error if user not found and letting the consumer check this is not good either.

Looking at the code handling the GetAuthInfoQuery seems like `models.ErrUserNotFound` is returned as error if no user is found. So I would suggest checking if `err == models.ErrUserNotFound` and if so debug log, otherwise error log. And keeping the `return nil, nil`.

billoley replied:
I understand checking models.ErrUserNotFound, logging at debug level, and still returning nil, nil in that case.
If there is an error that is not models.ErrUserNotFound, should I log that at error level and return the error (return nil, err)?

marefr replied:
> If there is an error that is not models.ErrUserNotFound, should I log that at error level and return the error (return nil, err)?

I was thinking log that at error level and return nil, nil to sort of mimic the old behavior. WDYT?

billoley replied:
The only place where the oauth token was being added in this way previously was in ds_proxy

if proxy.ds.JsonData != nil && proxy.ds.JsonData.Get("oauthPassThru").MustBool() {
	addOAuthPassThruAuth(proxy.ctx, req)
}

with the beginning of the addOAuthPassThruAuth function being:

func addOAuthPassThruAuth(c *models.ReqContext, req *http.Request) {
	authInfoQuery := &models.GetAuthInfoQuery{UserId: c.UserId}
	if err := bus.Dispatch(authInfoQuery); err != nil {
		logger.Error("Error fetching oauth information for user", "userid", c.UserId, "username", c.Login, "error", err)
		return
	}

To maintain the previous status quo, we have to *at some level* ignore *any* error that comes back from the GetAuthInfoQuery call to bus.Dispatch and carry on with sending the request without oauth. An error in that particular location would not stop the request.  This at least applies to the ds_proxy case since it existed previously.  To maintain backward compatibility and consistency, what you propose sounds fine to me.

billoley replied:
Committed that change to differentiate the log level

billoley replied:
Would it be possible to get this into the next 7.3.0 beta?

billoley replied:
Any idea why continuous-integration/drone/pr started failing on this PR?

marefr replied:
Sorry for slow response. Great. Will have a look. Unfortunately it may be too late to get this into v7.3 stable, but will consider/discuss it.

billoley committed:
Changes from pull request review

Changes from pull request review

billoley committed:
Differentiate log level for error in GetCurrentOAuthToken when getting token based on whether error is models.ErrUserNotFound

aknuds1 commented:
I would try merging with latest master to fix CI.

marefr committed:
fix ordering of imports

marefr committed:
Change GetCurrentOAuthToken to only return token

Makes a cleaner api to not returning an error and keep the logs
of errors in the oauthtoken package. One error created multiple
log messages earlier.
Decided to continue the request in both data source proxy and
backend plugin wrapper in case of an error and let the data source
handle the case of a missing token.

marefr commented:
Decided to change the `GetCurrentOAuthToken` to only return token. Makes a cleaner API to not returning an error and keep the logs of errors in the oauthtoken package for now. One error created multiple log messages earlier. Decided to continue the request in both data source proxy and backend plugin wrapper in case of an error and give the data source the responsibility of handling the case of a missing token/Authorization header. WDYT?

billoley commented:
> Decided to change the `GetCurrentOAuthToken` to only return token. Makes a cleaner API to not returning an error and keep the logs of errors in the oauthtoken package for now. One error created multiple log messages earlier. Decided to continue the request in both data source proxy and backend plugin wrapper in case of an error and give the data source the responsibility of handling the case of a missing token/Authorization header.
> 
> @billoley WDYT?

I agree that it's cleaner returning just the token (or nil) and handling all errors with logging inside the GetCurrentOAuthToken method.  I like it.

marefr reviewed (approved):
LGTM

marefr commented:
@billoley thank you for contributing to Grafana! Sorry it took some time to get this merged, but I really appreciate your patience and  attitude working together on this one.
